### ISSUE TITLE
lock contention caused by regexec

### ISSUE THREAD

**Issue by iksaif** [rsyslog/rsyslog#2759]
On a large setup, we found that regexec in https://github.com/rsyslog/rsyslog/blob/0d52ea8a918a5148c4be773fae5c64a63aea4c12/runtime/msg.c#L3925 could cause a lot of lock contention

```
34   Thread 0x7f837aecf700 (LWP 52339) "in:imjournal" 0x00007f8383c63a3d in poll () from /lib64/libc.so.6
  33   Thread 0x7f837aace700 (LWP 52340) "in:imudp" 0x00007f8383c6e923 in epoll_wait () from /lib64/libc.so.6
  32   Thread 0x7f837a6cd700 (LWP 52341) "in:imtcp" 0x00007f8384b7acf2 in pthread_cond_timedwait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
  31   Thread 0x7f837a2cc700 (LWP 52342) "in:impstats" 0x00007f8383c657a3 in select () from /lib64/libc.so.6
  30   Thread 0x7f8379ecb700 (LWP 52343) "in:imtcp" 0x00007f8384b7acf2 in pthread_cond_timedwait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
  29   Thread 0x7f8379aca700 (LWP 52344) "in:imtcp" 0x00007f8384b7acf2 in pthread_cond_timedwait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
  28   Thread 0x7f83796c9700 (LWP 52345) "in:imtcp" 0x00007f8384b7acf2 in pthread_cond_timedwait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
  27   Thread 0x7f83792c8700 (LWP 52346) "in:imtcp" 0x00007f8384b7a945 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
  26   Thread 0x7f8378cc1700 (LWP 52347) "rs:kestrel:Reg" 0x00007f8384b7d42d in __lll_lock_wait () from /lib64/libpthread.so.0
  25   Thread 0x7f8373fff700 (LWP 52348) "rs:omes-es5-log" 0x00007f8383c7beec in __lll_lock_wait_private () from /lib64/libc.so.6
  24   Thread 0x7f83737fe700 (LWP 52349) "rs:action 3 que" 0x00007f8384b7a945 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
  23   Thread 0x7f8372ffd700 (LWP 52350) "rs:action 4 que" 0x00007f8384b7a945 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
  22   Thread 0x7f83727fc700 (LWP 52354) "rs:kestrel:Reg" 0x00007f8384b7d42d in __lll_lock_wait () from /lib64/libpthread.so.0
  21   Thread 0x7f8371e68700 (LWP 52355) "rs:kestrel:Reg" 0x00007f8384b7d42d in __lll_lock_wait () from /lib64/libpthread.so.0
  20   Thread 0x7f8371667700 (LWP 52356) "rs:kestrel:Reg" 0x00007f8384b7d42d in __lll_lock_wait () from /lib64/libpthread.so.0
  19   Thread 0x7f8370a65700 (LWP 52357) "rs:kestrel:Reg" 0x00007f8384b7d42d in __lll_lock_wait () from /lib64/libpthread.so.0
  18   Thread 0x7f8323fff700 (LWP 52359) "rs:kestrel:Reg" 0x00007f8384b7d42d in __lll_lock_wait () from /lib64/libpthread.so.0
  17   Thread 0x7f83237fe700 (LWP 52360) "rs:kestrel:Reg" 0x00007f8384b7d42d in __lll_lock_wait () from /lib64/libpthread.so.0
  16   Thread 0x7f8322ffd700 (LWP 52361) "rs:kestrel:Reg" 0x00007f8384b7d42d in __lll_lock_wait () from /lib64/libpthread.so.0
  15   Thread 0x7f83227fc700 (LWP 52363) "rs:kestrel:Reg" 0x00007f8384b7d42d in __lll_lock_wait () from /lib64/libpthread.so.0
  14   Thread 0x7f8321ffb700 (LWP 52364) "rs:kestrel:Reg" 0x00007f8384b7d42d in __lll_lock_wait () from /lib64/libpthread.so.0
  13   Thread 0x7f8320ff9700 (LWP 52365) "rs:kestrel:Reg" 0x00007f8384b7d42d in __lll_lock_wait () from /lib64/libpthread.so.0
  12   Thread 0x7f82effff700 (LWP 52367) "rs:kestrel:Reg" 0x00007f8384b7d42d in __lll_lock_wait () from /lib64/libpthread.so.0
  11   Thread 0x7f82ef7fe700 (LWP 52368) "rs:kestrel:Reg" 0x00007f8384b7d42d in __lll_lock_wait () from /lib64/libpthread.so.0
  10   Thread 0x7f82eeffd700 (LWP 52369) "rs:kestrel:Reg" 0x00007f8384b7d42d in __lll_lock_wait () from /lib64/libpthread.so.0
  9    Thread 0x7f82ee7fc700 (LWP 52371) "rs:kestrel:Reg" 0x00007f8384b7d42d in __lll_lock_wait () from /lib64/libpthread.so.0
  8    Thread 0x7f83217fa700 (LWP 52402) "rs:omes-es5-log" 0x00007f8383c7beec in __lll_lock_wait_private () from /lib64/libc.so.6
  7    Thread 0x7f82ecffa700 (LWP 52412) "rs:main Q:Reg" 0x00007f8384b7a945 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
  6    Thread 0x7f82baefa700 (LWP 52495) "rs:omes-es5-log" 0x00007f8383c7beec in __lll_lock_wait_private () from /lib64/libc.so.6
  5    Thread 0x7f82ba6f9700 (LWP 52537) "rs:omes-es5-log" 0x00007f8383c46c1f in sift_states_backward () from /lib64/libc.so.6
  4    Thread 0x7f82b9ef8700 (LWP 52541) "rs:omes-es5-log" 0x00007f8383c7beec in __lll_lock_wait_private () from /lib64/libc.so.6
  3    Thread 0x7f82b96f7700 (LWP 52547) "rs:omes-es5-log" 0x00007f8383bf2182 in _int_free () from /lib64/libc.so.6
  2    Thread 0x7f82b8ef6700 (LWP 52555) "rs:omes-es5-log" 0x00007f8383c7beec in __lll_lock_wait_private () from /lib64/libc.so.6
* 1    Thread 0x7f83853e28c0 (LWP 52335) "rsyslogd" 0x00007f8383c657a3 in select () from /lib64/libc.so.6
```

```
(gdb) bt
#0  0x00007f8383c7beec in __lll_lock_wait_private () from /lib64/libc.so.6
#1  0x00007f8383c510c8 in _L_lock_60334 () from /lib64/libc.so.6
#2  0x00007f8383c50d1f in regexec@@GLIBC_2.3.4 () from /lib64/libc.so.6
#3  0x00005633ef950ad1 in MsgGetProp ()
#4  0x00005633ef97a444 in tplToString ()
#5  0x00005633ef93ae7c in doFuncCall ()
#6  0x00005633ef938cab in cnfexprEval ()
#7  0x00005633ef93b233 in doFuncCall ()
#8  0x00005633ef938cab in cnfexprEval ()
#9  0x00005633ef96d217 in scriptExec ()
#10 0x00005633ef96dafb in processBatch ()
#11 0x00005633ef925ebc in msgConsumer ()
#12 0x00005633ef96894a in ConsumerReg ()
#13 0x00005633ef965ba5 in wtiWorker ()
#14 0x00005633ef964c11 in wtpWorker ()
#15 0x00007f8384b76e25 in start_thread () from /lib64/libpthread.so.0
#16 0x00007f8383c6e34d in clone () from /lib64/libc.so.6
```

https://sourceware.org/bugzilla/show_bug.cgi?id=11159 advises to use one regexp_t object per thread.

Because it might be hard to do correctly for all callers, it might be worth considering a wrapper that does that semi-automatically using Thread-Local Storage

**Comment by rgerhards** [comment]
good find, albeit I really think the regexec implementation is buggy, as the man page says it is MT-save...

**Comment by iksaif** [comment]
Yep, but https://sourceware.org/bugzilla/show_bug.cgi?id=11159#c6 doesn't give much hope of this being fixed :)

**Comment by rgerhards** [comment]
yup, saw that - still there is a bug. I think at least the man page needs to be fixed (what also means glibc is not POSIX compliant here). If I only I would think it would not be wasted effort to open a bug tracker against glibc ;-)

**Comment by unbrice** [comment]
Just passing by but "MT-safe" does not mean "lock-free". I indeed doubt the glibc authors would agree this is a bug.

**Comment by rgerhards** [comment]
@unbrice you are right, my bad. That happens if you try to do too much in parallel - I misread "locking contention" with "deadlock"... 

So in conclusion this issue is not a bug report but a feature request.

**Comment by iksaif** [comment]
I'll try working on that today and see what I can get done.

**Comment by iksaif** [comment]
Work in progress on https://github.com/iksaif/rsyslog

**Comment by iksaif** [comment]
Finally got a way to test this:

Without the patch (notice that cpu is capped at ~100%):

{code}
$ ps  aux | grep rsyslog
c.chary   6997  108  1.2 1277632 396836 pts/0  SNl+ 17:03   0:23 ./local/sbin/rsyslogd -i /home/cchary/dev/rsyslog/rsyslogd.pid -n -f rsyslog.conf

(gdb) info threads
  Id   Target Id         Frame 
* 1    Thread 0x7f38c6006c40 (LWP 6997) "rsyslogd" 0x00007f38c4a6103f in __GI___select (nfds=1, readfds=0x0, writefds=0x0, exceptfds=0x0, timeout=0x7ffcb02f8b50) at ../sysdeps/unix/sysv/linux/select.c:41
  2    Thread 0x7f38c28f7700 (LWP 6998) "in:imudp" __lll_lock_wait () at ../sysdeps/unix/sysv/linux/x86_64/lowlevellock.S:135
  3    Thread 0x7f38c24f6700 (LWP 6999) "in:imtcp" 0x00007f38c4a6bbb7 in epoll_wait (epfd=9, events=events@entry=0x7f38c24f4ab0, maxevents=128, timeout=timeout@entry=-1)
    at ../sysdeps/unix/sysv/linux/epoll_wait.c:30
  4    Thread 0x7f38c20f5700 (LWP 7000) "in:imtcp" 0x00007f38c59ee9f3 in futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x7f38c35a8e10 <wrkrInfo+48>)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:88
  5    Thread 0x7f38c1cf4700 (LWP 7001) "in:imtcp" 0x00007f38c59ee9f3 in futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x7f38c35a8e78 <wrkrInfo+152>)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:88
  6    Thread 0x7f38c18f3700 (LWP 7002) "in:imtcp" 0x00007f38c59ee9f3 in futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x7f38c35a8ee0 <wrkrInfo+256>)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:88
  7    Thread 0x7f38c14f2700 (LWP 7003) "in:imtcp" 0x00007f38c59ee9f3 in futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x7f38c35a8f48 <wrkrInfo+360>)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:88
  8    Thread 0x7f38c10f1700 (LWP 7004) "rs:kestrel:Reg" 0x000055a768cc31b8 in jsonAddVal (
    pSrc=pSrc@entry=0x7f38b824c550 "/delivery/api/search?apiKey=764937a9-7afb-4cf1-8631-d8a5edca8f36&puserid=01639F2F64E905014DE50B214F0F8540&platform=desktop&environmentHost=https%3A%2F%2Fwww.target.com&keyword=crib+bedding+set&taxonom"..., buflen=1341, dst=dst@entry=0x7f38c10f0750, escapeAll=1) at msg.c:3330
  9    Thread 0x7f38c08f0700 (LWP 7005) "rs:kestrel:Reg" __GI___libc_free (mem=0x7f38ac4254e0) at malloc.c:3101
  10   Thread 0x7f38b3fff700 (LWP 7006) "rs:kestrel:Reg" __lll_lock_wait_private () at ../sysdeps/unix/sysv/linux/x86_64/lowlevellock.S:95
  11   Thread 0x7f38b37fe700 (LWP 7007) "rs:kestrel:Reg" re_node_set_add_intersect (dest=dest@entry=0x7f38b37fd480, src1=src1@entry=0x7f38b800dbb8, src2=src2@entry=0x7f38b800ee58) at regex_internal.c:1110
  12   Thread 0x7f38b2ffd700 (LWP 7008) "rs:kestrel:Reg" __lll_lock_wait_private () at ../sysdeps/unix/sysv/linux/x86_64/lowlevellock.S:95
  13   Thread 0x7f38b27fc700 (LWP 7009) "rs:kestrel:Reg" __lll_lock_wait_private () at ../sysdeps/unix/sysv/linux/x86_64/lowlevellock.S:95
  14   Thread 0x7f38b1ffb700 (LWP 7010) "rs:kestrel:Reg" __lll_lock_wait_private () at ../sysdeps/unix/sysv/linux/x86_64/lowlevellock.S:95
  15   Thread 0x7f38b17fa700 (LWP 7011) "rs:kestrel:Reg" __lll_lock_wait_private () at ../sysdeps/unix/sysv/linux/x86_64/lowlevellock.S:95
  16   Thread 0x7f38b0ff9700 (LWP 7012) "rs:kestrel:Reg" 0x00007f38c4a3ddf8 in re_node_set_compare (set2=0x7f38b0ff8340, set1=0x7f38b80124a8) at regex_internal.c:1355
  17   Thread 0x7f388ffff700 (LWP 7013) "rs:kestrel:Reg" __lll_lock_wait_private () at ../sysdeps/unix/sysv/linux/x86_64/lowlevellock.S:95
(gdb) thread 15
[Switching to thread 15 (Thread 0x7f38b17fa700 (LWP 7011))]
#0  __lll_lock_wait_private () at ../sysdeps/unix/sysv/linux/x86_64/lowlevellock.S:95
95      ../sysdeps/unix/sysv/linux/x86_64/lowlevellock.S: No such file or directory.
(gdb) bt
#0  __lll_lock_wait_private () at ../sysdeps/unix/sysv/linux/x86_64/lowlevellock.S:95
#1  0x00007f38c4a4b6f7 in __GI___regexec (eflags=<optimized out>, pmatch=<optimized out>, nmatch=<optimized out>, string=<optimized out>, preg=<optimized out>) at regexec.c:215
#2  re_exec (s=<optimized out>) at regexec.c:563
#3  0x0000000000000000 in ?? ()
(gdb) 
{code}

With the patch
{code}
$ ps  aux | grep rsyslog
c.chary   8454  236  0.0 949952 26348 pts/0    SNl+ 17:05   0:56 ./local/sbin/rsyslogd -i /home/cchary/dev/rsyslog/rsyslogd.pid -n -f rsyslog.conf

(venv) (master) c.chary@workstation:~/dev/defcon-criteo$ sudo gdb -p 8454
GNU gdb (Ubuntu 8.1-0ubuntu3) 8.1.0.20180409-git
Copyright (C) 2018 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word".
Attaching to process 8454
[New LWP 8455]
[New LWP 8456]
[New LWP 8457]
[New LWP 8458]
[New LWP 8459]
[New LWP 8460]
[New LWP 8461]
[New LWP 8462]
[New LWP 8463]
[New LWP 8464]
[New LWP 8465]
[New LWP 8466]
[New LWP 8467]
[New LWP 8468]
[New LWP 8469]
[New LWP 8470]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
0x00007f2f11b8203f in __GI___select (nfds=1, readfds=0x0, writefds=0x0, exceptfds=0x0, timeout=0x7ffdbcdb9f20) at ../sysdeps/unix/sysv/linux/select.c:41
41      ../sysdeps/unix/sysv/linux/select.c: No such file or directory.
(gdb) info threads
  Id   Target Id         Frame 
* 1    Thread 0x7f2f13127c40 (LWP 8454) "rsyslogd" 0x00007f2f11b8203f in __GI___select (nfds=1, readfds=0x0, writefds=0x0, exceptfds=0x0, timeout=0x7ffdbcdb9f20) at ../sysdeps/unix/sysv/linux/select.c:41
  2    Thread 0x7f2f0fa18700 (LWP 8455) "in:imudp" ParseMsg (pMsg=0x7f2f08114880) at parser.c:655
  3    Thread 0x7f2f0f617700 (LWP 8456) "in:imtcp" 0x00007f2f11b8cbb7 in epoll_wait (epfd=10, events=events@entry=0x7f2f0f615ab0, maxevents=128, timeout=timeout@entry=-1)
    at ../sysdeps/unix/sysv/linux/epoll_wait.c:30
  4    Thread 0x7f2f0f216700 (LWP 8457) "rs:kestrel:Reg" 0x00007f2f11b5eb22 in re_acquire_state (err=err@entry=0x7f2f0f215240, dfa=dfa@entry=0x7f2f04003020, nodes=nodes@entry=0x7f2f0f215300)
    at regex_internal.c:1468
  5    Thread 0x7f2f0ea15700 (LWP 8458) "in:imtcp" 0x00007f2f12b0f9f3 in futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x7f2f106c9e10 <wrkrInfo+48>)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:88
  6    Thread 0x7f2f0e614700 (LWP 8459) "in:imtcp" 0x00007f2f12b0f9f3 in futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x7f2f106c9e78 <wrkrInfo+152>)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:88
  7    Thread 0x7f2f0e213700 (LWP 8460) "in:imtcp" 0x00007f2f12b0f9f3 in futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x7f2f106c9ee0 <wrkrInfo+256>)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:88
  8    Thread 0x7f2f0de12700 (LWP 8461) "in:imtcp" 0x00007f2f12b0f9f3 in futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x7f2f106c9f48 <wrkrInfo+360>)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:88
  9    Thread 0x7f2f0da11700 (LWP 8462) "rs:kestrel:Reg" 0x00007f2f12b0ff85 in futex_abstimed_wait_cancelable (private=<optimized out>, abstime=0x7f2f0da10bb0, expected=0, futex_word=0x55a7791dce6c)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:205
  10   Thread 0x7f2f0d210700 (LWP 8463) "rs:kestrel:Reg" 0x00007f2f12b0ff85 in futex_abstimed_wait_cancelable (private=<optimized out>, abstime=0x7f2f0d20fbb0, expected=0, futex_word=0x55a7791ff1a8)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:205
  11   Thread 0x7f2f0ca0f700 (LWP 8464) "rs:kestrel:Reg" 0x00007f2f12b0ff85 in futex_abstimed_wait_cancelable (private=<optimized out>, abstime=0x7f2f0ca0ebb0, expected=0, futex_word=0x55a7792016f8)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:205
  12   Thread 0x7f2ef7fff700 (LWP 8465) "rs:kestrel:Reg" re_node_set_contains (set=0x7f2eec00eb48, set=0x7f2eec00eb48, elem=19) at regex_internal.c:1381
  13   Thread 0x7f2ef77fe700 (LWP 8466) "rs:kestrel:Reg" 0x00007f2f12b0ff85 in futex_abstimed_wait_cancelable (private=<optimized out>, abstime=0x7f2ef77fdbb0, expected=0, futex_word=0x55a77920619c)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:205
  14   Thread 0x7f2ef6ffd700 (LWP 8467) "rs:kestrel:Reg" 0x00007f2f12b0ff85 in futex_abstimed_wait_cancelable (private=<optimized out>, abstime=0x7f2ef6ffcbb0, expected=0, futex_word=0x55a7792086ec)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:205
  15   Thread 0x7f2ef67fc700 (LWP 8468) "rs:kestrel:Reg" calc_state_hash (nodes=0x7f2ef67fb300, nodes=0x7f2ef67fb300, context=0) at regex_internal.c:1451
  16   Thread 0x7f2ef5ffb700 (LWP 8469) "rs:kestrel:Reg" 0x00007f2f12b0ff85 in futex_abstimed_wait_cancelable (private=<optimized out>, abstime=0x7f2ef5ffabb0, expected=0, futex_word=0x55a77920d188)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:205
  17   Thread 0x7f2ef57fa700 (LWP 8470) "rs:kestrel:Reg" 0x00007f2f12b0ff85 in futex_abstimed_wait_cancelable (private=<optimized out>, abstime=0x7f2ef57f9bb0, expected=0, futex_word=0x55a77920f6dc)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:205
{code}

More CPU is used with the patch and the threads are almost all waiting for new work to do (which shows that there is no more contention on regexec)

**Comment by rgerhards** [comment]
Apologies for the very late follow-up. This may no longer help the original reporter, but I am recording the resolution for people finding this through search while working through the backlog. We do not close reports merely because they are old.

Commit 62161ae32d ("regexp.c: reduce lock contention when using glibc") implemented per-thread regular-expression state on glibc and removed the global regexec bottleneck described here. That work was delivered through PR #2786.

Closing as fixed. If profiling a current release shows renewed regex lock contention, please open a new issue with the profile and the parser/template workload.
